### PR TITLE
Copy Minimal API docs into separate sections

### PR DESCRIPTION
**Pull request by halter73**
- Also expand MapGroup descriptions
- Includes and replaces #27424

[Internal review URL](https://review.learn.microsoft.com/en-us/aspnet/core/fundamentals/minimal-apis?view=aspnetcore-7.0&branch=pr-en-us-27458)

### CODE REVIEW

File: aspnetcore/includes/route-groups.md
Comment (captainsafia) [inline aspnetcore/includes/route-groups.md:19]: Maybe be more explicit here about the scenario of global filters? Feels more compelling than an example with the `all` tags.

File: aspnetcore/release-notes/aspnetcore-7.0.md
Comment (captainsafia) [inline aspnetcore/release-notes/aspnetcore-7.0.md:189]: It doesn't make sense to include the entirety of the route groups doc in the release notes...

File: aspnetcore/toc.yml
Comment (captainsafia) [inline aspnetcore/toc.yml:703]: Is there a more scenario-based title we could use here. Something like "Constructing minimal application" or what not...

File: aspnetcore/fundamentals/routing.md
Comment (captainsafia) [inline aspnetcore/fundamentals/routing.md:972]: This is only relevant for library authors correct? For example, when is 

> This gives complete control over running group conventions and the final metadata on the grouped endpoints.

important?

File: aspnetcore/fundamentals/minimal-apis/webapplication.md
Comment (captainsafia) [inline aspnetcore/fundamentals/minimal-apis/webapplication.md:11]: Ditto the commentary below on the TOC. This title doesn't feel particularly helpful...

File: aspnetcore/fundamentals/minimal-apis/test-min-api.md
Comment (captainsafia) [inline aspnetcore/fundamentals/minimal-apis/test-min-api.md:4]: @Rick-Anderson I thought we preferred `apps` over `applications` due to shortness?

File: aspnetcore/fundamentals/minimal-apis/route-handlers.md
Comment (captainsafia) [inline aspnetcore/fundamentals/minimal-apis/route-handlers.md:13]:
```suggestion
A configured `WebApplication` supports `Map{Verb}` and <xref:Microsoft.AspNetCore.Builder.EndpointRouteBuilderExtensions.MapMethods%2A>:
```

Clarify what "Verb" means here?

File: aspnetcore/fundamentals/minimal-apis/includes/route-handlers.md
Comment (captainsafia) [inline aspnetcore/fundamentals/minimal-apis/includes/route-handlers.md:81]: A link instead of an include makes more sense here...

File: aspnetcore/release-notes/aspnetcore-7.0.md
Comment (halter73) [reply aspnetcore/release-notes/aspnetcore-7.0.md:189]: Do you really think so? This seems smaller than the "OpenAPI improvements for minimal APIs" section and I consider it a similarly big feature. And it's a feature that needs to be demonstrated in conjunction of something else like endpoint filters.

That said, it'd be very easy to revert the changes in this file if people generally agree this is too much detail.

File: aspnetcore/includes/route-groups.md
Comment (halter73) [reply aspnetcore/includes/route-groups.md:19]: What about `.WithOpenApi()`? We already show adding a filter to a group just above. I'll add a suggestion to the relevant line.

File: aspnetcore/includes/route-groups.md
Comment (halter73) [inline aspnetcore/includes/route-groups.md:22]: > Maybe be more explicit here about the scenario of global filters? Feels more compelling than an example with the all tags.

```suggestion
var all = app.MapGroup("").WithOpenApi();
```

File: aspnetcore/toc.yml
Comment (halter73) [reply aspnetcore/toc.yml:703]:
```suggestion
      - name: Constructing Minimal API apps
```

I slightly edited this to match the naming convention of the other sections. I still think "WebApplication and WebApplicationBuilder" is more obvious and clearly defined. "Constructing minimal applications" could literally describe creating the entire application including the routes and middleware.

Does anyone else have an opinion on this?

File: aspnetcore/fundamentals/routing.md
Comment (halter73) [reply aspnetcore/fundamentals/routing.md:972]: > This is only relevant for library authors correct?

Correct. This is in the `## Guidance for library authors` section.

> For example, when is [...] important?

What do you think about?

```suggestion
**CONSIDER** implementing <xref:Microsoft.AspNetCore.Routing.EndpointDataSource.GetGroupedEndpoints%2A>. This gives complete control over running group conventions and the final metadata on the grouped endpoints. For example, this allows custom `EndpointDataSource` implementations to run [endpoint filters](xref:fundamentals/minimal-apis/min-api-filters) added to groups.
```

The guidance we're currently giving for implementing `EndpointDataSource` at all is just above this.

> **CONSIDER** writing your own <xref:Microsoft.AspNetCore.Routing.EndpointDataSource>. `EndpointDataSource` is the low-level primitive for declaring and updating a collection of endpoints. `EndpointDataSource` is a powerful API used by controllers and Razor Pages.

So, these recommendations aren't intended to be super in-depth. I still haven't seen a custom `EndpointDataSource`, so I don't think it's worthwhile to spend much time expanding more on thius.

File: aspnetcore/fundamentals/minimal-apis/includes/route-handlers.md
Comment (halter73) [reply aspnetcore/fundamentals/minimal-apis/includes/route-handlers.md:81]: We could do this:

```suggestion
See <xref:fundamentals/routing#route-groups>.
```

There are a few issues with this approach though. One is that this link is likely to break in the future since it includes a fragment. More fundamentally, I think people would expect to see this content with the other minimal route handling concepts. Linking out to the complicated routing doc isn't going to be good for beginners.

I think route groups should be considered on close to equal footing with something like "Wildcard and catch all routes" and is more important than "Route constraints" both of which are above this in the doc.

**Comment by JamesNK** [comment]
The titles in the TOC and in the doc should be switched.

TOC title: Route handlers
Doc title: Route handlers in ASP.NET Core Minimal API

In the TOC, the link is already nested under Minimal API. And in the page, it should have the full name for people who are linked to it, and for search engine indexing.

image

Repeat for all.

File: aspnetcore/fundamentals/minimal-apis/route-handlers.md
Comment (halter73) [reply aspnetcore/fundamentals/minimal-apis/route-handlers.md:13]: This isn't new, but I agree it can be improved. It's unfortunate this is one of the few places where I was not able to deduplicate. It would be really easy to fix here but not on the cheat sheet if you hadn't just worked this. Still we've deduped most of the content. What do we think about this?

```suggestion
A configured `WebApplication` supports `Map{Verb}` and <xref:Microsoft.AspNetCore.Builder.EndpointRouteBuilderExtensions.MapMethods%2A> where `{Verb}` is a camel-cased HTTP method like `Get`, `Post`, `Put` or `Delete`:
```

**Comment by halter73** [comment]
> The titles in the TOC and in the doc should be switched.

I completely agree with this 💯% @JamesNK, but I was just sticking with the exiting convention. If we want to switch these, I think we should do that in a follow up PR. What do you think @Rick-Anderson and @tdykstra?

File: aspnetcore/toc.yml
Comment (halter73) [inline aspnetcore/toc.yml:701]:
```suggestion
      - name: Minimal APIs cheat sheet
```

File: aspnetcore/fundamentals/minimal-apis/7.0-samples/todo-group/Program.cs
Comment (JamesNK) [inline aspnetcore/fundamentals/minimal-apis/7.0-samples/todo-group/Program.cs:15]: Is this supported in .NET 6? If not, then may need to have separate code samples for .NET 6 and .NET 7.

**Review by Rick-Anderson — CHANGES_REQUESTED** [review]
All your changes are in the `:::moniker range="= aspnetcore-6.0"` section, was that your intention?  Once you get approval of the .NET 6 version, then update the .NET 7 section?

cc @tdykstra

File: aspnetcore/fundamentals/minimal-apis/includes/webapplication.md
Comment (Rick-Anderson) [inline aspnetcore/fundamentals/minimal-apis/includes/webapplication.md:17]:
```suggestion
When a web app is created with Visual Studio or `dotnet new`, a `Properties/launchSettings.json` file is created that specifies the ports the app responds to. In the port setting samples that follow, running the app from Visual Studio returns an error dialog `Unable to connect to web server 'AppName'`.  Visual Studio returns an error because it's expecting the port specified in `Properties/launchSettings.json`, but the app is using the port specified by `app.Run("http://localhost:3000")`. Run the following port changing samples from the command line.
```

File: aspnetcore/fundamentals/minimal-apis/includes/webapplication.md
Comment (Rick-Anderson) [inline aspnetcore/fundamentals/minimal-apis/includes/webapplication.md:15]: This seem super niche. How often is it useful the specify the port?  If it's rare, this shouldn't be at the top.

**Comment by Rick-Anderson** [comment]
> > The titles in the TOC and in the doc should be switched.
> 
> I completely agree with this 💯% @JamesNK, but I was just sticking with the existing convention. If we want to switch these, I think we should do that in a follow up PR. What do you think @Rick-Anderson and @tdykstra?

We frequently fix the TOC in a follow-up PR.

File: aspnetcore/toc.yml
Comment (tdykstra) [reply aspnetcore/toc.yml:703]: > I still think "WebApplication and WebApplicationBuilder" is more obvious and clearly defined. "Constructing minimal applications" could literally describe creating the entire application including the routes and middleware.

I agree. I'm not fond of "WebApplication and WebApplicationBuilder" as it doesn't describe the content very clearly -- but I think it's better than "Constructing minimal applications" since that could be the title for the entire cheat sheet. WA and WAB could possibly be split up into several H2 sections with titles more clearly describing their contents, like "Working with ports", "Working with certificates" etc. But that's a task for another time, and I would keep WA and WAB for now.

File: aspnetcore/fundamentals/minimal-apis.md
Comment (tdykstra) [reply aspnetcore/fundamentals/minimal-apis.md:2]: Consider what we have for System.Text.Json for comparison -- 

* An [Overview](https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/overview) that gives a high-level description of features and when/why to use it. 
* A main how-to doc like the single cheat sheet for Minimal APIs - "[How to serialize and deserialize JSON in .NET](https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/how-to)". (It got too big and we extracted parts of it into other articles, but it started out as the one STJ cheat sheet.)

So expanding on @Rick-Anderson's suggestion, we could name this "How to create and configure Minimal API apps."

File: aspnetcore/fundamentals/minimal-apis/7.0-samples/todo-group/Program.cs
Comment (halter73) [reply aspnetcore/fundamentals/minimal-apis/7.0-samples/todo-group/Program.cs:15]: This already is in the 7.0-samples directory. Groups are not supported by .NET 6 either. This is only referenced from .NET 7 and greater docs.

File: aspnetcore/fundamentals/minimal-apis.md
Comment (halter73) [reply aspnetcore/fundamentals/minimal-apis.md:2]: I thought we agreed to call it a "Cheat sheet" instead of an "overview" because this is far more in-depth and comprehensive. Is that not what we agreed to?

I'm fine with reverting this, but I don't want to go back and forth on this. I want to get this merged ASAP because this is a major refactoring that is likely to conflict with work others are doing in this area. @JeremyLikness @samsp-msft @davidfowl

File: aspnetcore/fundamentals/minimal-apis.md
Comment (halter73) [reply aspnetcore/fundamentals/minimal-apis.md:313]: I was just fixing an `or` to `of` typo. I do not want to waste time in this PR review rewriting all the existing minimal API docs. That will take forever.

Can you please file a follow up issue or PR fixing these kind of semantic issues with the existing docs?

File: aspnetcore/fundamentals/minimal-apis.md
Comment (halter73) [reply aspnetcore/fundamentals/minimal-apis.md:2]: So the suggested options so far are:

```suggestion
title: Minimal APIs overview
```

```suggestion
title: Minimal APIs quick overview
```

```suggestion
title: Minimal APIs quick reference
```

```suggestion
title: Minimal APIs how to
```

```suggestion
title: How to create and configure Minimal API apps
```

```suggestion
title: Minimal APIs how to
```

I prefer "cheat sheet", but I'm happy to switch it to whatever else has consensus. I just want an answer for which one I should switch to. I'll update all the other references to the title to match.

File: aspnetcore/fundamentals/minimal-apis.md
Comment (Rick-Anderson) [reply aspnetcore/fundamentals/minimal-apis.md:313]: #27475

**Comment by halter73** [comment]
> All your changes are in the `:::moniker range="= aspnetcore-6.0"` section, was that your intention? Once you get approval of the .NET 6 version, then update the .NET 7 section?

Very few of my changes are in the `:::moniker range="= aspnetcore-6.0"` section. Almost all of my changes where I extracted content out into `[!INCLUDE`s are after line 779 in minimal-apis.md which is the aspnetcore-7.0 section. 

The changes before that line are only to change the title, update the overview, include #27424 because it would merge conflict, and fix the smallest typos.

I did miss was updating the overview in the aspnetcore-7.0 section though. I'm fixing that now. This kind of mistake is so easy to make which is why I hate copy and pasting large swaths content into multiple places ☹️

File: aspnetcore/fundamentals/minimal-apis.md
Comment (halter73) [reply aspnetcore/fundamentals/minimal-apis.md:17]: I can revert this. I'm just waiting for our decision on the title to see if we can have it match whatever we go for with that. I think @JeremyLikness is planning to rewrite this intro anyway.

File: aspnetcore/fundamentals/minimal-apis.md
Comment (halter73) [reply aspnetcore/fundamentals/minimal-apis.md:2]: Assuming "cheat sheet" is an absolute no go, I think "quick reference" is my second favorite option.

Comment (Rick-Anderson) [reply aspnetcore/fundamentals/minimal-apis.md:2]: > Assuming "cheat sheet" is an absolute no go, I think "quick reference" is my second favorite option.

Definitely not an absolute `no go`, but slang is strongly discouraged. 

`quick reference` is a great alternative that readily MTs and doesn't loose much for those familiar with English slang.

Comment (davidfowl) [reply aspnetcore/fundamentals/minimal-apis.md:2]: I like quick reference

**Comment by Rick-Anderson** [comment]
> I did miss updating the ~overview~ intro in the aspnetcore-7.0 section of minimal-apis.md though. I just pushed a fix for that. This kind of mistake is so easy to make which is why I hate copy and pasting large swaths content into multiple places ☹️

Maybe @guardrex  could help me get the new file based versioning working. That would fix the copy/paste where are you in the file problem and the broken internal links problem.

**Comment by guardrex** [comment]
I couldn't get it fully working, so I had to drop it. I hope Microsoft figures it out and lets me know how to set it up.

**Review by rafikiassumani-msft — COMMENTED** [review]
@Rick-Anderson , @halter73 Can we add a navigation menu for the cheat sheet? I think a menu with links will make all the sections easily discoverable. I know the argument for ctrl + f, but a navigation menu makes it easier for people to go directly to the section they want reference for.

**Review by rafikiassumani-msft — COMMENTED** [review]
@halter73 and @Rick-Anderson Should each fragment/section be a submenu item for better discoverability?

**Comment by Rick-Anderson** [comment]
> @Rick-Anderson , @halter73 Can we add a navigation menu for the cheat sheet? I think a menu with links will make all the sections easily discoverable. I know the argument for ctrl + f, but a navigation menu makes it easier for people to go directly to the section they want reference for.

You need a very wide browser window to get the H2 menu at the right and that only shows the top 4 H2's
image

Show more:

image


> @Rick-Anderson , @halter73 Can we add a navigation menu for the cheat sheet? I think a menu with links will make all the sections easily discoverable.

Do you want an unorder list at the top with an in page anchor to each H2 and H3 or do you want to promote all the H3's to H2's and use the automatically generated TOC on the right?

cc @tdykstra

**Comment by rafikiassumani-msft** [comment]
@Rick-Anderson Unordered list at the top should be enough. We could even categorize later. I will create an issue.

**Comment by halter73** [comment]
I removed all reference to "cheat sheet" and replaced it with "quick reference". I plan to open a PR immediately after closing this to switch titles in the TOC and the doc themselves as @JamesNK suggested.

Is there anything remaining that's blocking us from merging this PR and doing the rest as follow ups?

**Comment by Rick-Anderson** [comment]
> I removed all reference to "cheat sheet" and replaced it with "quick reference". I plan to open a PR immediately after closing this to switch titles in the TOC and the doc themselves as @JamesNK suggested.
> 
> Is there anything remaining that's blocking us from merging this PR and doing the rest as follow ups?

Merge it. I can do the TOC if you like.

File: aspnetcore/fundamentals/minimal-apis/test-min-api.md
Comment (Rick-Anderson) [reply aspnetcore/fundamentals/minimal-apis/test-min-api.md:4]: Oops

**Pull request merged**
